### PR TITLE
NIAD-110: outbound deduction

### PR DESCRIPTION
**Pull request by mattd-kainos**

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapper.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+@Component
+public class DeductionDateMapper implements FromFhirToEdifactMapper<DateTimePeriod> {
+
+    public DateTimePeriod map(Parameters parameters) {
+        return new DateTimePeriod(getDeductionDate(parameters), DateTimePeriod.TypeAndFormat.DEDUCTION_DATE);
+    }
+
+    private Instant getDeductionDate(Parameters parameters) {
+        String dateAsString = ParametersExtension.extractValue(parameters, ParameterNames.DATE_OF_DEDUCTION);
+        return LocalDate.parse(dateAsString).atStartOfDay(ZoneOffset.UTC).toInstant();
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapper.java]: Date of entry/exit uses:
`return localDate.atStartOfDay(TimestampService.UKZone).toInstant();`
which one is right?

**Review by grzegorz-robak — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapper.java (hunk 1 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapper.java]: good question I just copied this from acceptance. Since they are date it might be better to use LocalDate all the way through instead of assigning them a nominal time.

Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapper.java]: So `return Instant.from(localDate);` ?

Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapper.java]: I think you can't get `Instant` without time part

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDate.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+@Getter @Setter @AllArgsConstructor @EqualsAndHashCode(callSuper = false) @ToString
+public class DeductionDate extends Segment{
+
+    public static final String KEY = "DTM";
+    private static final String TYPE_CODE = "961";
+    private static final String FORMAT_CODE = "102";
+    private static final DateTimeFormatter FORMAT = DateTimeFormatter.ofPattern("yyyMMdd");
+
+    private LocalDate date;
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDate.java]: this could be `private final @NonNull LocalDate date` and class only needs `@RequiredArgsConstructor`

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.NonNull;
+import org.apache.commons.lang3.EnumUtils;
+import uk.nhs.digital.nhsconnect.nhais.model.edifact.message.EdifactValidationException;
+
+import java.util.Objects;
+
+@EqualsAndHashCode(callSuper = false)
+@Builder
+@Data
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java]: Make `code` final and only `@RequiredArgsConstructor` will be needed

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        parameters.addParameter()
+            .setName(ParameterNames.DATE_OF_DEDUCTION)
+            .setValue(new StringType("2020-06-05"));
+
+        var deductionDateMapper = new DeductionDateMapper();
+        DeductionDate deductionDate = deductionDateMapper.map(parameters);
+
+        var expectedDeductionDate = new DeductionDate(LocalDate.parse("2020-06-05"));
+
+        assertEquals(expectedDeductionDate, deductionDate);
```
Comment (grzegorz-robak) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapperTest.java]: Should use AssertJ

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionReasonCodeMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            .setValue(new StringType("20"));
+
+        var mapper = new DeductionReasonCodeMapper();
+        DeductionReasonCode deductionReasonCode = mapper.map(parameters);
+
+        var expectedDeductionReasonCode = DeductionReasonCode.builder()
+            .code("20")
+            .build();
+
+        assertEquals(expectedDeductionReasonCode, deductionReasonCode);
```
Comment (grzegorz-robak) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionReasonCodeMapperTest.java]: AssertJ

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDateTest.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.time.LocalDate;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class DeductionDateTest {
+
+    @Test
+    public void When_deductionDateToEdifact_Then_edifactFormattedCorrectly() {
+        DeductionDate deductionDate = new DeductionDate(LocalDate.parse("2005-01-15"));
```
Comment (grzegorz-robak) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDateTest.java:13]: add test for `new DeductionDate(null)`

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCodeTest.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Test
+    public void When_MappingToEdifact_Then_ReturnCorrectString() {
+        var expectedValue = "GIS+11:ZZZ'";
+
+        var deductionReasonCode = DeductionReasonCode.builder()
+            .code("11")
+            .build();
+
+        assertEquals(expectedValue, deductionReasonCode.toEdifact());
```
Comment (grzegorz-robak) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCodeTest.java]: assertJ

**Review by grzegorz-robak — CHANGES_REQUESTED** [review]

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCodeTest.java (hunk 7 above)
Comment (mattd-kainos) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCodeTest.java]: done

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDateTest.java (hunk 6 above)
Comment (mattd-kainos) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDateTest.java:13]: used NonNull annotation instead

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionReasonCodeMapperTest.java (hunk 5 above)
Comment (mattd-kainos) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionReasonCodeMapperTest.java]: done

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapperTest.java (hunk 4 above)
Comment (mattd-kainos) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/DeductionDateMapperTest.java]: done

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java (hunk 3 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java]: done

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDate.java (hunk 2 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionDate.java]: done

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        return code.concat(ZZZ_SUFFIX);
+    }
+
+    @Override
+    protected void validateStateful() throws EdifactValidationException {
+    }
+
+    @Override
+    public void preValidate() throws EdifactValidationException {
+        if (Objects.isNull(code) || code.isBlank()) {
```
Comment (bartek-sarul) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java]: `StringUtils.isBlank` should cover both checks

**Review by bartek-sarul — DISMISSED** [review]

**Review by grzegorz-robak — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java (hunk 8 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/DeductionReasonCode.java]: removed null check since NonNull annotation is used

**Review by grzegorz-robak — DISMISSED** [review]

**Review by grzegorz-robak — APPROVED** [review]

**Review by bartek-sarul — APPROVED** [review]
